README.md
# HCL Model

![PyTest](https://github.com/khrapovs/hcl-model/actions/workflows/pytest.yaml/badge.svg)
![Docs](https://github.com/khrapovs/hcl-model/actions/workflows/docs.yaml/badge.svg)
[![!pypi](https://img.shields.io/pypi/v/hcl-model)](https://pypi.org/project/hcl-model/)
[![codecov](https://codecov.io/gh/khrapovs/hcl-model/branch/main/graph/badge.svg?token=KC0XT6R18H)](https://codecov.io/gh/khrapovs/hcl-model)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![!python-versions](https://img.shields.io/pypi/pyversions/hcl-model)](https://pypi.org/project/hcl-model/)

Simple time series forecasting based on multiple linear regression.

## Documentation

Full documentations hosted on GitHub pages: [khrapovs.github.io/hcl-model](https://khrapovs.github.io/hcl-model/).

## Installation

```shell
pip install hcl-model
```

## Contribute

Create a virtual environment and activate it
```shell
python -m venv venv
source venv/bin/activate
```
Install the development packages
```shell
pip install -e .[dev]
```
and use pre-commit to make sure that your code is formatted using [black](https://github.com/PyCQA/isort) and [isort](https://pycqa.github.io/isort/index.html) automatically:
```shell
pre-commit install
```
Run tests:
```shell
pip install -e .[test]
pytest
```
Build documentation:
```shell
pip install -e .[doc]
mkdocs build
```
or use
```shell
mkdocs serve
```
if you prefer a live, self-refreshing, documentation.


CHANGELOG.md
# Changelog

## 0.6.2 (2022-02-22)
### Changed
- Refactor `hcl_model.utils.decayed_weights.decayed_weights` and allow `None` as input for `full_weight_obs` and `downweight_order`. See more details in the docstring.

## 0.6.1 (2022-02-22)
### Changed
- Rename argument `full_weight_weeks` to `full_weight_obs` in `hcl_model.transformers.truncate.TruncateTransformer`. Truncation is now calculated in terms of number of observations instead of datetime indexing to accommodate numpy arrays.

## 0.6.0 (2022-02-21)
### Changed
- Break down `CalendarTransformer` into `sklearn` compatible transformers: `AddAutomaticSeasonalDummies`, `AddHolidayDummies`, `AddHolidayTriangles`, and `AddPeriodicSplines`.

## 0.5.6 (2022-02-21)
### Added
- Trivial `inverse_transform` method that returns the input to `TruncateTransformer`, `TargetOutlierCorrectionTransformer`, and `TargetStructuralBreakCorrectionTransformer`. This addition allows using these transformers in `sklearn.compose.TransformedTargetRegressor` as `tranformer` argument for trivial inverse transform of pipeline predictions.
- Argument `weights` in `HandCraftedLinearModel`. It is a function that is applied to endogenous variable and passed to WLS for fitting.
### Changed
- Expose fitted attributes with trailing underscore: `x_train_`, `y_train_`, `fit_results_`.
- Expose model parameters, e.g. `endog_transform` and `exog_transform` in `H andCraftedLinearModel`.

## 0.5.5 (2022-02-18)
### Changed
- If `num_steps` is `None`, get it from number of `X` observations in `hcl_model.model_base.ModelBase.predict`.

## 0.5.4 (2022-02-17)
### Changed
- Allow `np.ndarray` type for `TargetOutlierCorrectionTransformer`, and `TargetStructuralBreakCorrectionTransformer`. 

## 0.5.3 (2022-02-17)
### Changed
- Change parameter order in `hcl_model.model_base.ModelBase.fit` and `.predict`. This is important if the estimator is used in `sklearn.pipeline.Pipeline` object.
- Allow `np.ndarray` for `y` input of the `fit` method. In this case its index is inherited from `X`. If `X` is `None`, then `TypeError` is risen. 

## 0.5.2 (2022-02-17)
### Changed
- Store all transformer parameters as class attributes with the same name during initialization. 

## 0.5.1 (2022-02-15)
### Changed
- Convert previously private method to public: `hcl_model.model_base.ModelBase.get_parameters`.

## 0.5.0 (2022-02-08)
### Added
- `sklearn` compliant transformer `TargetOutlierCorrectionTransformer` in `hcl_model.transformers.outlier_correction`. Read documentation in `docs/data_preprocessing.md`.
- `sklearn` compliant transformer `TargetStructuralBreakCorrectionTransformer` in `hcl_model.transformers.structural_breaks`. It uses [`ruptures` package](https://github.com/deepcharles/ruptures/) to correct structural breaks in `X` data.
- `sklearn` compliant transformer `EstimatorToTransformer` in `hcl_model.transformers.estimator_to_transformer`. Effectively, it replaces `predict` method with `transform` so that the object is usable as an intermediate step in `Pipeline`.
- `sklearn` compliant transformer `TruncateTransformer` in `hcl_model.transformers.truncate`. Leave only a specific number of past weeks in the data.
- `sklearn` compliant transformer `SeriesToFrameTransformer` in `hcl_model.transformers.series_to_frame`. It simply applies `.to_frame()` to a `pd.Series`.
- A replacement for `sklearn.pipeline.FeatureUnion`: `DatetimeIndexedFeatureUnion` in `hcl_model.transformers.feature_union`.
- A replacement for `sklearn.pipeline.make_union`: `make_union_of_datetime_indexed_features` in `hcl_model.transformers.feature_union`.
- Move utility functions from `utils.py` to a separate module `utils` with `.py` file per function.
#### Changed
- Move `CalendarTransformer` to `hcl_model.transformers.calendar`.
- Move `CalendarReader` to `hcl_model.utils.calendar_reader.CalendarReader`.

## 0.4.0 (2022-01-26)
- Rename arguments: `exog` into `X`, and `endog` into `y`. This is done to comply with [`sklearn`](https://scikit-learn.org/stable/developers/develop.html) general interface.
- Now `.fit` should be fed only with past data, and `.predict` only with future data.

## 0.3.9 (2021-05-19)
- Allow passing arbitrary transformations to `HandCraftedLinearModel` that return multi-column dataframes.

## 0.3.8 (2021-05-17)
- Use datetime index of exogenous and endogenous variables. 

## 0.3.7 (2021-03-09)
- Remove upper constraint on `pandas` version.

## 0.3.6 (2021-01-04)
- Remove version restriction on `skyfield`.
- Limit `pandas` version by 1.1.5 pending incompatibility fix `patsy` 0.5.1.

## 0.3.5 (2020-12-10)
- Use `median_abs_deviation` instead of deprecated `median_absolute_deviation`. This requires `scipy>=1.5`.
- Use `dates.isocalendar().week` instead of deprecated `dates.week`. This requires `pandas>=1.1`.
- Silence warning about default `dtype` in empty `Series`. Now `dtype=float`.

## 0.3.4 (2020-10-21)
- Change the way seasonal outliers are detected. Instead of looking at percentage changes, the focus now is at the absolute deviation from the exponentially weighted moving average.

## 0.3.3 (2020-07-24)
- Limit the version of `skyfield` package to 1.22 due to an error `AttributeError: module 'datetime' has no attribute 'combine'`.

## 0.3.2 (2020-06-05)
- Require `workalendar` version to be greater than 10.0.0 due to a rename of `IsoRegistry.get_calendar_class()` into `IsoRegistry.get()`.
 
## 0.3.1 (2020-01-27)
- Fix the bug in automatic dummy detection. Instead of taking weeks from sorted normalized series of changes, they were taken from some other intermediate result. 

## 0.3 (2020-01-10)
- Add several utility functions that help construct exogenous regressors based on a calendar. These functions are accessible through `construct_calendar_exogenous`. 
- Add utility function to create decayed weights for weighted OLS estimation.

## 0.2.1 (2020-01-21)
- Add parameter dictionary to model summary output.

## 0.2 (2020-01-10)

src/hcl_model/calendar_transformer.py
from typing import Iterable

import pandas as pd
from patsy import dmatrix
from scipy.stats import median_abs_deviation

from hcl_model.calendar_reader import CalendarReader
from hcl_model.labels import LabelsCommon, LabelsExog


class CalendarTransformer:
    """
    Allows to easily add a variety of calendar-based columns to the exogenous regressor matrix.
    """

    lbl = LabelsCommon()
    lbl_exog = LabelsExog()

    lbl_weekly_freq = "7 D"
    lbl_auto_dummy = "automatic_dummy_{}"

    def __init__(self):
        self._cal_reader = CalendarReader()

    def add_periodic_splines(
        self,
        df: pd.DataFrame,
        input_level: str = "isoweek",
        degrees_of_freedom: int = 4,
    ) -> pd.DataFrame:
        """
        Add periodic spline scores to the dataframe df.

        df must have a datetime column as index. This method adds periodic splines
        scores to the dataframe by using its index as input.

        Typical output:

        .. code-block:: python

            df = pd.DataFrame({'x1': np.random.normal(0, 1, 53 * 3)},
                               index=pd.date_range(start='2015-01-01', freq='7 D', periods=53 * 3))

            cal = CalendarTransformer()
            cal.add_periodic_splines(df).head()

                              x1  spline_dim_1  spline_dim_2  spline_dim_3  spline_dim_4
            2015-01-01 -0.301179     -0.439735     -0.438071     -0.438801     -0.437531
            2015-01-08  0.301794     -0.343647     -0.460494     -0.409252     -0.498250
            2015-01-15 -0.431848     -0.218113     -0.474556     -0.374233     -0.526180
            2015-01-22  2.669445     -0.071386     -0.479289     -0.335690     -0.526172
            2015-01-29  0.220345      0.088282     -0.473725     -0.295568     -0.503074

        """
        data = df.copy(deep=True)
        data[self.lbl_exog.lbl_calendar_numeric] = self._get_numeric_periodicity(
            data.index, input_level=input_level
        ).astype(float)

        # the 'cc' part in the string of the dmatrix call is responsible for periodic splines.
        formula = f"cc({self.lbl_exog.lbl_calendar_numeric}, df={degrees_of_freedom}, constraints='center') - 1"
        calendar_matrix = dmatrix(formula, data=data, return_type="dataframe")

        # rename columns for convenience
        cols = calendar_matrix.columns
        calendar_matrix = calendar_matrix.rename(
            columns={
                col: f"{self.lbl_exog.lbl_spline_dim}_{num + 1}"
                for num, col in enumerate(cols)
            }
        )
        data[calendar_matrix.columns] = calendar_matrix

        return data.drop(self.lbl_exog.lbl_calendar_numeric, axis=1)

    def _get_numeric_periodicity(
        self, dates: pd.DatetimeIndex, input_level: str = "isoweek"
    ) -> pd.Series:
        """
        Add a column of numeric scores between 0 and 1 that represent annual periodicity.
        Input levels 'isoweek' and 'month' supported currently. For any other input we fall
        back to day of the year. The effect of using week or month is that all dates within one
        week or month get the same score while fall-back routine assigns each day of year same score.

        dates must have datetime data type.
        """
        if input_level == self.lbl.isoweek:
            return dates.isocalendar().week / 53
        elif input_level == self.lbl.month:
            return dates.month / 12
        elif input_level == self.lbl.day:
            return dates.isocalendar().day / 366
        else:
            raise ValueError(f"Unknown input_level argument: {input_level}")

    def add_holiday_triangles(
        self,
        df: pd.DataFrame,
        holiday_name: str,
        country_code: str,
        triangle_name: str = "",
        lh_factors: Iterable[float] = (1.0, 2.0, 3.0),
        rh_factors: Iterable[float] = (2.0, 1.0),
        by: str = None,
        merge_factors: bool = False,
    ) -> pd.DataFrame:
        """
        Adds triangular calendar regressors to the dataframe df. Index of df must be sorted and of date type.
        If there are additional grouping columns, then this can be set with the `by` argument (see pd.merge_asof).

        The user must pass two list, specifying the factors values that will be used.
        lh_factors = [..., factor one week before the holiday, factor on the week of the holiday]
        rh_factors = [factor on the week of the holiday, factor one week after the holiday, ...]

        If `merge_factors` = False, then a single numeric column of holiday scores is added. If not, two columns
        are added, representing effects before (including at) and after the holiday, respectively.

        Typical output:

        .. code-block:: python

            df = pd.DataFrame({'x1': np.random.normal(0, 1, 53 * 3)},
                               index=pd.date_range(start='2015-01-01', freq='7 D', periods=53 * 3))

            cal = CalendarTransformer()
            cal.add_holiday_triangles(df, 'New year', 'DE').head()

                              x1  _lh_factor  _rh_factor
            2015-01-01 -0.301179         3.0         2.0
            2015-01-08  0.301794         0.0         1.0
            2015-01-15 -0.431848         0.0         0.0
            2015-01-22  2.669445         0.0         0.0
            2015-01-29  0.220345         0.0         0.0
        """
        lbl_lh = "{}_lh_factor".format(triangle_name)
        lbl_rh = "{}_rh_factor".format(triangle_name)

        dates = self._read_holiday_dates(holiday_name, country_code)

        lh_df = list()
        for i, factor in enumerate(reversed([0.0] + list(lh_factors))):

            if merge_factors and i == 0:
                # if the user required the lh and rh factors to be merged in one, take only half value for the
                # day shared between lh and rh
                factor = factor * 0.5

            lh_df.append(
                pd.DataFrame(
                    {
                        self.lbl.date: dates[self.lbl.date]
                        - pd.Timedelta("{} D".format(7 * i)),
                        lbl_lh: [factor] * len(dates),
                    }
                )
            )
        lh_df = pd.concat(lh_df).sort_values([self.lbl.date], ascending=True)

        # - build right-hand holiday factors
        rh_df = list()
        for i, factor in enumerate(list(rh_factors) + [0.0]):

            if merge_factors and i == 0:
                # if the user required the lh and rh factors to be merged in one, take only half value for the
                # day shared between lh and rh
                factor = factor * 0.5

            rh_df.append(
                pd.DataFrame(
                    {
                        self.lbl.date: dates[self.lbl.date]
                        + pd.Timedelta("{} D".format(7 * i)),
                        lbl_rh: [factor] * len(dates),
                    }
                )
            )
        rh_df = pd.concat(rh_df).sort_values([self.lbl.date], ascending=True)

        # - generate new DataFrame with all holidays factors
        edf = (
            pd.merge(left=lh_df, right=rh_df, how="outer", on=self.lbl.date)
            .fillna(0)
            .reset_index(drop=True)
        )

        # - if the user required the lh and rh factors to be merged in one, merge them by summing the two columns
        # NOTE: the lh and rh factors in the case of merged factor are built to make the sum of the two columns
        # producing the merged factor
        if merge_factors:
            edf["{}_factor".format(triangle_name)] = edf[lbl_lh] + edf[lbl_rh]
            edf = edf.drop(columns=[lbl_lh, lbl_rh])

        # - insert all other missing days in the calendar, with factor 0
        edf = (
            edf.set_index(self.lbl.date)
            .sort_index()
            .reindex(
                pd.date_range(
                    start="{}-01-01".format(dates[self.lbl.date].dt.year.min()),
                    end="{}-12-31".format(dates[self.lbl.date].dt.year.max()),
                    freq="D",
                ),
                method="nearest",
            )
            .reset_index()
            .rename(columns={"index": self.lbl.date})
        )

        # - remove the time from the date column
        edf[self.lbl.date] = pd.to_datetime(edf[self.lbl.date].dt.date)

        return pd.merge_asof(
            df,
            edf.set_index(self.lbl.date),
            left_index=True,
            right_index=True,
            by=by,
            tolerance=pd.Timedelta(self.lbl_weekly_freq),
        )

    def add_holiday_dummies(
        self,
        df: pd.DataFrame,
        holiday_name: str,
        country_code: str,
        dummy_name: str = None,
        add_td: str = None,
        lags: Iterable[int] = (-2, -1, 0, 1, 2),
        by: str = None,
    ) -> pd.DataFrame:
        """
        Adds calendar effects to dataframe df with either 0 or 1 in each entry. 1 indicates that the corresponding
        date is close to the holiday that is considered.

        Inputs:
            df: dataframe with datetime index.
            holiday_name: name of the holiday. See CalendarReader.
            country_code: calendar's country code. See CalendarReader.
            dummy_name: optional name for renaming the dummy columns.
            add_td: adds to each holiday a fixed time delta. See add_cc_dummies as a typical example.
            lags: list of integers; add forward and backward looking lags of the dummy.
            by: name of group column if df should be grouped by that column when the merge_asof is applied.
        Output:
            returns df extended by dummy columns.

        Typical output:

        .. code-block:: python

            df = pd.DataFrame({'x1': np.random.normal(0, 1, 53 * 3)},
                               index=pd.date_range(start='2015-01-01', freq='7 D', periods=53 * 3))

            cal = CalendarTransformer()
            cal.add_holiday_dummies(df, 'Chinese New Year', 'TW').iloc[7:12, :]

                              x1  is_holiday  ...  is_holiday_lag_1  is_holiday_lag_2
            2015-02-19 -1.127138         0.0  ...               0.0               0.0
            2015-02-26  0.303294         1.0  ...               0.0               0.0
            2015-03-05 -0.398241         0.0  ...               1.0               0.0
            2015-03-12  1.281731         0.0  ...               0.0               1.0
            2015-03-19  2.303174         0.0  ...               0.0               0.0
        """
        dummy = self._get_holiday_dummy(holiday_name, country_code, add_td)

        if dummy_name is not None:
            dummy = dummy.rename(columns={self.lbl_exog.lbl_is_holiday: dummy_name})

        return self._add_dummy_lags_and_merge(df, dummy, lags, by)

    def _get_holiday_dummy(
        self, holiday_name: str, country_code: str, add_td: str = None
    ) -> pd.DataFrame:
        """
        Any holiday included in the input dataframe can be translated to dummy column.
        Use e.g. `holiday_name` = 'Chinese New Year' and `country_code` = 'CN'.
        """
        dates = self._read_holiday_dates(holiday_name, country_code).assign(
            **{self.lbl_exog.lbl_is_holiday: 1}
        )
        if add_td is not None:
            dates[self.lbl.date] = dates[self.lbl.date] + pd.Timedelta(add_td)

        # TODO: this is a very dirty quick fix!!!
        dates[self.lbl.date] -= pd.Timedelta("7 D")

        return pd.merge_asof(
            pd.DataFrame({self.lbl.date: self._get_weekly_date_range()}),
            dates,
            on=self.lbl.date,
            allow_exact_matches=False,
            tolerance=pd.Timedelta(self.lbl_weekly_freq),
        ).fillna(0)

    def _add_dummy_lags_and_merge(
        self, df: pd.DataFrame, dummy, lags: Iterable[int], by: str
    ) -> pd.DataFrame:
        """
        Add optional forward and backward looking lags and merge to dataframe.
        """
        dummy_name = dummy.columns[-1]
        for lag in lags:
            lag_name = "{}_lag_{}".format(dummy_name, str(lag).replace("-", "neg"))
            dummy[lag_name] = dummy[dummy_name].shift(lag)

        # Add 0 to lags to preserve original dummy and rename:
        dummy = dummy.drop(dummy_name, axis=1).rename(
            columns={"{}_lag_0".format(dummy_name): dummy_name}
        )

        return pd.merge_asof(
            df,
            dummy.set_index(self.lbl.date),
            left_index=True,
            right_index=True,
            by=by,
            tolerance=pd.Timedelta(self.lbl_weekly_freq),
        )

    def _get_weekly_date_range(
        self, start: str = "2005-01-03", periods: int = 53 * 25
    ) -> pd.DatetimeIndex:
        return pd.date_range(start=start, periods=periods, freq=self.lbl_weekly_freq)

    def _read_holiday_dates(
        self,
        holiday_name: str,
        country_code: str,
        from_year: int = 2000,
        to_year: int = 2030,
    ) -> pd.DataFrame:
        return (
            self._cal_reader.get_holidays(
                holiday_name, country_code, from_year, to_year
            )
            .reset_index()
            .filter(items=[self.lbl.date])
        )

    @classmethod
    def add_automatic_seasonal_dummies(
        cls,
        df: pd.DataFrame,
        var_name: str,
        lim_num_dummies: int = 5,
        threshold: float = 3,
    ) -> pd.DataFrame:
        """Add automatic seasonal dummies.

        Outliers among weekly percentage changes are detected by normalizing and comparing with a certain threshold.
        Only weekly frequency is supported.

        :param df: original data
        :param var_name: the name of the variable to model and forecast
        :param lim_num_dummies: limit on the number of seasonal dummies
        :param threshold: quantile cutoff for "irregular" time series changes
        :return: original data with new columns corresponding to seasonal dummies
        """
        freq = pd.infer_freq(df.index)
        if freq[0] != "W":
            raise RuntimeError(
                "Only weekly data is supported. Frequency detected: {}".format(freq)
            )

        lbl_diff = "diff"
        lbl_week_number = "week_number"
        data = df.copy()
        data[lbl_week_number] = data.index.map(lambda x: x.isocalendar()[1])
        data[lbl_diff] = data[var_name] - data[var_name].ewm(com=10).mean()
        data[lbl_diff] = (data[lbl_diff] / data[lbl_diff].std()).abs()
        mean_abs_diff = (
            data.iloc[10:].groupby(lbl_week_number)[lbl_diff].mean().dropna()
        )
        normalized = (
            (mean_abs_diff - mean_abs_diff.median()).abs()
            / median_abs_deviation(mean_abs_diff, scale=1 / 1.4826 ** 2)
        ).sort_values(ascending=False)
        weeks = normalized.loc[normalized > threshold].index[:lim_num_dummies]
        for week in weeks:
            data[cls.lbl_auto_dummy.format(week)] = 0.0
            data.loc[
                data.index.map(lambda x: x.isocalendar()[1]) == week,
                cls.lbl_auto_dummy.format(week),
            ] = 1
        return data.drop([lbl_week_number, lbl_diff], axis=1)
